code/modules/mob/living/bot/ed209bot.dm

```diff
@@ -19,6 +29,201 @@
+	var/list/command_buffer = list()
+	var/list/known_commands = list("stay", "stop", "arrest", "detain", "follow", "patrol")
+	var/emote_hear = "states"
+	move_to_delay = 3
+
+/mob/living/bot/secbot/ed209/Initialize()
+	..()
+	if(!short_name)
+		short_name = name
+
+/mob/living/bot/secbot/ed209/hear_say(var/message, var/verb = "says", var/datum/language/language = null, var/alt_name = "", var/italics = 0, var/mob/speaker = null, var/sound/speech_sound, var/sound_vol)
+	if(thinking_enabled && !stat && has_ui_access(speaker))
+		command_buffer.Add(speaker)
+		command_buffer.Add(lowertext(html_decode(message)))
+	return 0
+
+/mob/living/bot/secbot/ed209/hear_radio(var/message, var/verb="says", var/datum/language/language=null, var/part_a, var/part_b, var/part_c, var/mob/speaker = null, var/hard_to_hear = 0)
+	if(thinking_enabled && !stat && has_ui_access(speaker))
+		command_buffer.Add(speaker)
+		command_buffer.Add(lowertext(html_decode(message)))
+	return 0
+
+/mob/living/bot/secbot/ed209/think()
+	while(command_buffer.len > 0)
+		var/mob/speaker = command_buffer[1]
+		var/text = command_buffer[2]
+		var/filtered_name = lowertext(html_decode(name))
+		var/filtered_short = lowertext(html_decode(short_name))
+		if(dd_hasprefix(text,filtered_name))
+			var/substring = copytext(text,length(filtered_name)+1) //get rid of the name.
+			listen(speaker,substring)
+		else if(dd_hasprefix(text,filtered_short))
+			var/substring = copytext(text,length(filtered_short)+1) //get rid of the name.
+			listen(speaker,substring)
+		command_buffer.Remove(command_buffer[1],command_buffer[2])
+	..()
+	switch(mode)
+		if(SECBOT_FOLLOW)
+			follow_target()
+		if(SECBOT_STOP)
+			commanded_stop()
+
+/mob/living/bot/secbot/ed209/on_think_disabled()
+	..()
+	command_buffer.Cut()
+
+/mob/living/bot/secbot/ed209/proc/follow_target()
+	if(!target)
+		return
+	if(target in view(7, src))
+		walk_to(src,target,1,move_to_delay)
+
+/mob/living/bot/secbot/ed209/proc/commanded_stop() //basically a proc that runs whenever we are asked to stay put. Probably going to remain unused.
+	return
+
+/mob/living/bot/secbot/ed209/proc/listen(var/mob/speaker, var/text)
+	for(var/command in known_commands)
+		if(findtext(text,command))
+			switch(command)
+				if("stay")
+					if(stay_command(speaker,text)) //find a valid command? Stop. Dont try and find more.
+						break
+				if("stop")
+					if(stop_command(speaker,text))
+						break
+				if("arrest")
+					if(arrest_command(speaker,text))
+						break
+				if("detain")
+					if(arrest_command(speaker,text))
+						break
+				if("follow")
+					if(follow_command(speaker,text))
+						break
+				if("patrol")
+					if(patrol_command(speaker, text))
+						break
+
+	return 1
+
+//returns a list of everybody we wanna do stuff with.
+/mob/living/bot/secbot/ed209/proc/get_target_by_name(var/text)
+	var/list/possible_targets = hearers(src,10)
+	for(var/mob/M in possible_targets)
+		var/found = 0
+		if(findtext(text, "[M]"))
+			found = 1
```

Comment:
Standard coding practice is just ```return M``` if you found something

Reply:
Yeah, you are right. That is a weird way. I copied code from commanded mobs and did not revise it entirely.

Reply:
It's fine. I copy existing code all the time and usually it ends up being bad code.